[Bug][P1] Add temperature and weather_components arguments handling

deerhunter4 opened the pull request:
argument_parser.py was modified. Two arguments added: 'temperature' and weather_components.
Additionally corresponding tests for those arguments were added to check_args() function.

deerhunter4 committed:
[Bug][P1] Add temperature and weather_components arguments handling
argument_parser.py was modified. Two arguments added:
'temperature' and weather_components.
Additionally corresponding tests for those arguments
were added to check_args() function.

KrzysiekJodlowski commented on argument_parser.py:
```diff
@@ -6,6 +6,10 @@
+TEMP_ARG = "-temp"
+AVALAIBLE_TEMP = ['C', 'F']
+WEATHER_ARG = '-weather_components'
+AVALAIBLE_WEATHER = ['wind', 'pressure']
```
Only `wind` and  `pressure`? We also have at least `temperature` and `rain` mentioned in the Feature Document.

deerhunter4 replied:
@KrzysiekJodlowski Yes, but temperature and rain are always shown, you cannot "unchose" them. However, I agree that the list's name might be confusing so I can change it to something like ADDITIONAL_WEATHER_PARAM.

KrzysiekJodlowski replied:
@deerhunter4 got it, yes you are right, `temperature` and `rain` will always be considered, according to the FD, sorry I missed that somehow. And thank you for updating the variable name, I'd only suggest naming it `ADDITIONAL_WEATHER_PARAMS` since there is more than one of them.

deerhunter4 replied:
@KrzysiekJodlowski not up to date, check belowe.

KrzysiekJodlowski commented on argument_parser.py:
```diff
@@ -60,5 +72,14 @@ def get_parameters():
                         The correct format should be YYYY-MM-DD""",
                         default=current_date)
 
+    parser.add_argument('-temp', metavar='temperature', type=str, choices=['C', 'F'],
```
By the FD, this one is named `-temperature`, not `-temp`. When I'm trying to execute with the extended name:
```
> python .\main.py -location "New York" -date 2024-06-17 -temperature "F" -weather_components "pressure"
2024-06-16
usage: BetterWeather [-h] [-location location] [-fp forecast_period] [-date date] [-temp temperature] [-weather_components weather_components]
> python .\main.py -location "New York" -date 2024-06-17 -temperature "C" -weather_components "pressure"
2024-06-16
usage: BetterWeather [-h] [-location location] [-fp forecast_period] [-date date] [-temp temperature] [-weather_components weather_components]
BetterWeather: error: unrecognized arguments: -temperature C
```
Also, for the `choices`, why we don't reuse the already existing variable here?

deerhunter4 replied:
@KrzysiekJodlowski Solved.

KrzysiekJodlowski commented on argument_parser.py:
```diff
@@ -60,5 +72,14 @@ def get_parameters():
+                        scale (default: C).""", default='C')
+
+    parser.add_argument('-weather_components', metavar='weather_components',
+                        type=str, choices=['wind', 'pressure'], help="""Select
```
The `type=str` makes this parameter a one-to-pick option. The documentation states:
>Additional weather components to choose from, e.g.: ‘wind’, ‘pressure’, …

..which makes the impression we should be able to pick one or more components here, therefore I believe this should be a list parameter. And again, for the `choices`, why we don't reuse the already existing variable here?

deerhunter4 replied:
@KrzysiekJodlowski 
It is not so trivial. Changing the type from 'str' to 'list' does not help when you have an argument with choices. You would need to have lists to choose from. So what I did, for now, is leave type= str and add one option to choices = ['wind', 'pressure', 'both']. Other options would be creating two arguments: pressure and wind, with options yes/no.

KrzysiekJodlowski replied:
@deerhunter4 got it, yes, as we discussed separately this can be solved in two ways:
- either making it a list parameter and checking it only with a custom code, leaving out the `choices` since it accepts only one option from the list
- making those separate params, although that means more code and also updating the FD accordingly

I'm fine with both options and also with an alternative to have them handled in a later phase. Please think about it and let me know, though, at first glance, I think we should have them and have them as a list.

deerhunter4 replied:
@KrzysiekJodlowski 
So I finally decided to make separate parameters from weather_components. And because it is now a 'flag' parameter we do not need any extra tests for them (at least I think so). I also corrected README.md file and FD file.
Additionally, I want to point out that actually there was a possibility to make parameter in type list. Check 'append' for acion [https://docs.python.org/3/library/argparse.html#action](url) . Still, I think that making two separated parameters is better idem.

KrzysiekJodlowski replied:
@deerhunter4 got it, thanks for updating this, this was one of the options to go with, and using flag params simplifies that part - good idea!

KrzysiekJodlowski reviewed (changes requested):
@deerhunter4 nice to see you're reusing the already-existing patterns here (well, in the vast majority)! But please address all the comments I left. Also:
- since you're introducing new parameters here, shouldn't we handle them in the requests?
- each change impacting the app behavior should be documented in the README file, and because we have the argument pool growing I'd suggest making a list with short descriptions of those (and if the param is accepted but for some reason not handled by the app, this information should also be stated)

deerhunter4 commented:
@KrzysiekJodlowski 
```
since you're introducing new parameters here, shouldn't we handle them in the requests?
```
Yes, we should, but please let's do it when the whole app will work for basic (default) parameters.
```
each change impacting the app behavior should be documented in the README file, and because we have the argument pool
growing I'd suggest making a list with short descriptions of those (and if the param is accepted but for some
reason not handled by the app, this information should also be stated)
```
Done. README.md updated.

I addressed all the comments. I hope that you will be satisfied :)

deerhunter4 committed:
[Bug][P1] Add temperature and weather_components arguments - correction
README.md and argument_parser.py were updated.

KrzysiekJodlowski commented:
> Yes, we should, but please let's do it when the whole app will work for basic (default) parameters.

Got it, ok, makes sense to do it step by step. Just make sure to mention that in the README, and, *please prepare an update suggestion in the FD to mention that some of the params handling will be implemented in a later phase* (pay attention to the fact that it should be a SUGGESTION, not a direct edit, this way I'll be able to review it before applying).

> Done. README.md updated.
> I addressed all the comments. I hope that you will be satisfied :)

Thanks! Looks good, though I have some minor comments that I'll place directly in the code changes.

KrzysiekJodlowski commented on README.md:
```diff
@@ -2,9 +2,18 @@
+**Parameters**  
+*Required:*  
+-location - Name of the city/town in English. E.g. 'Madrid', 'Valencia', 'Cracovia’
 
-**Requirements:**
+*Optional:*  
+-forecast_period - Forecast period, with six options to choose from: '24h', '48h', '72h', '7days', '10days', '14days'. (default: 24h)  
+-date - Date for which weather forecast will be presented or will start from, e.g.: '2024-04-01'. (default: current date)  
+-temperature - Two temperature scales to choose from: Celsius [C] and Fahrenheit [F]. (default: C) **!!!not supported for now!!!**  
+-weather_components - Additional weather components to choose from: 'wind', 'pressure', 'both'. **!!!not supported for now!!!**
```
Please update two things:
- make the params name formatted with backticks (`) to distinguish them from the other text
- make the notifications about not being supported in a less screamy way, it looks a bit "unprofessional"😛 (e.g. without the/so many exclamation marks and in braces)

deerhunter4 replied:
@KrzysiekJodlowski Done.

deerhunter4 committed:
[Bug][P1] Add temperature and weather_components arguments - update
Weather_components parameter was replaced by two separate
parameters: pressure and wind.
README.md file has been corrected accordingly to the above changes.

KrzysiekJodlowski commented:
@deerhunter4 please note that - according to the notification - "This branch has conflicts that must be resolved".  **You need to fix that before the merging**.

deerhunter4 committed:
Merge branch 'main' into bug.p1.add_temperature_weather_components
